Pokergameapp step4

okstring opened the pull request:
- 이번 step에서는 플레이어가 카드를 보이면 안된다는 생각에 protocol을 계속해서 봐꿔봤던 것 같습니다. 
 - 또한 이전 리뷰에서도 언급하셨듯이 extension이 크지 않아서 안에서 구현해봤습니다 

   해보면서 extension에서 궁금했던 건 제 평소 생각에 뭔가 해당 class, struct가 좀 더 깔끔해보이고(CustomStringConvertible, CaseIterable 채택을 extension에) 부가기능(?)의 성격이 띈 속성, 메소드들은 extension으로 선언하려 노력하고 있었습니다. 사실 제가 판단하는 부가기능의 기준도 모호하고 짧은 코드는 그냥 해당 class에 붙여도 되는건지 궁금합니다!

okstring committed:
수정사항 반영 - Dealar 추가, shuffle 직접 구현

okstring committed:
팩토리 메소드, Dealer, Player 구현

 - Dealr에는 start, 캐릭터 구현, 카드 share, print 나눠 구현

okstring committed:
Dealer도 고심 끝에 playable 채택

okstring committed:
CustomStringConvertible class 안에서 구현, README.md 작성

okstring committed:
Modify stack view init(그룹 리뷰 참고)

godrm commented:
> * 해보면서 extension에서 궁금했던 건 제 평소 생각에 뭔가 해당 class, struct가 좀 더 깔끔해보이고(CustomStringConvertible, CaseIterable 채택을 extension에) 부가기능(?)의 성격이 띈 속성, 메소드들은 extension으로 선언하려 노력하고 있었습니다. 사실 제가 판단하는 부가기능의 기준도 모호하고 짧은 코드는 그냥 해당 class에 붙여도 되는건지 궁금합니다!

그 기준은 스스로 정하는 겁니다. 그리고 저처럼 코드를 보는 다른 개발자를 납득하고 설득할 수 있으면 됩니다. 
어떤 경우에 extension을 쓰는 게 좋을까요?

godrm commented on PokerGameApp/PokerGameApp/Card.swift:
```diff
@@ -1,17 +1,36 @@
+        }
+        return cards
+    }
+}
```
protocol + extension 도 사용하는 의도가 있을 때 사용하는 게 좋습니다. 그냥 써보고 싶다고 쓰는게 아니라 그걸 해야하는 이유, 이득이 있다거나 해야 하는거죠. 
예를 들어 CardCreatable 프로토콜 대신에 CardCreatable 구조체를 그냥 선언하는 것과 어떤 차이가 있을까요? 프로토콜을 선언하면 뭐가 더 편리해지나요? 이런 질문을 생각해보고 고민해보고 판단하는 게 소프트웨어 엔지니어의 고민이 되어야 합니다.

godrm commented on PokerGameApp/PokerGameApp/Card.swift:
```diff
@@ -23,44 +42,21 @@ class Card {
         }
+        
+        var description: String {
+            return rankName
```
rankName 이란 computed property가 꼭 필요한가요? description과 동일한 기능이면 여기에 구현해도 되는거 아닐까요?

godrm commented on PokerGameApp/PokerGameApp/CardDeck.swift:
```diff
@@ -21,6 +22,20 @@ struct CardDeck {
+            shuffledElements[index] = shuffledElements[randomIndex]
+            shuffledElements[randomIndex] = temp
+        }
+        return shuffledElements
```
좋은 시도입니다. 아마도 대부분의 경우는 Array.shuffle() 이나 shuffled()를 그대로 쓰겠죠.
하지만 우리는 익숙해지는 과정에서 직접 구현해보는 것도 도움이 됩니다.

godrm commented on PokerGameApp/PokerGameApp/Dealer.swift:
```diff
@@ -0,0 +1,67 @@
+        case fiveCardStud = 5
+    }
+    
+    var name = "딜러"
```
프로토콜에 get으로 선언했으면 let name으로 선언해도 되지 않을까요?

godrm commented on PokerGameApp/PokerGameApp/Dealer.swift:
```diff
@@ -0,0 +1,67 @@
+        return cards
+    }
+    
+    public func start(numberOfPlayer: UInt, stud: CardStud) {
```
CardStud 처럼 numberOfPlayer 도 UInt가 아니라 타입을 선언하는 게 어떨까요? 
100, 1000명도 play가 가능한가요?
물론 아래 예외처리를 해주기는 하지만 타입이 그것을 표현해주면 더 좋습니다.

godrm commented on PokerGameApp/PokerGameApp/Dealer.swift:
```diff
@@ -0,0 +1,67 @@
+    }
+    
+    private func shareCards(players: [Playable], count: Int) {
+        for _ in 1...count {
```
CardStud를 타입으로 선언했으면 count를 CardStud로 쓰는 게 좋습니다. 
Int로 사용할 꺼면 타입을 선언해놓고 제대로 활용하지 못하는 겁니다.

godrm commented on PokerGameApp/PokerGameApp/Dealer.swift:
```diff
@@ -0,0 +1,67 @@
+protocol Playable {
+    var name: String { get }
+    var playResult: [Card] { get }
+    func getCard(_ card: Card)
```
프로토콜 선언에서 어색한 부분이 있습니다. 
- playResult 배열을 통째로 넘겨주는 게 좋은 선택일까요?
- getCard()인데 return 타입이 없네요. 어떻게 get하는거죠?

godrm commented on PokerGameApp/PokerGameApp/Player.swift:
```diff
@@ -0,0 +1,24 @@
+    var name: String
+    var cards = [Card]()
+    var playResult: [Card] {
+        return cards
```
cards가 public 이면 외부에서 배열을 바꿀 수 있는건가요?

godrm commented on README.md:
```diff
@@ -1,138 +1,17 @@
-[참고](https://seizze.github.io/2019/12/20/iOS-메모리-뜯어보기,-메모리-이슈-디버깅하기,-메모리-릭-찾기.html)
-
-[참고2](https://blog.canapio.com/130)
 
```
학습한 내용들이 좋아보이는 데 다 지워졌네요?

godrm commented:
위에 피드백 확인해보고 수정할 게 없으면 다음단계로 넘어가겠습니다. 확인해주세요.

okstring commented:
헉 고민할 거리를 많이 주셔서 감사합니다. 리뷰해주신 부분 더욱 깊게 그리고 많이 생각해보고 수정이 대폭 필요해보입니다. Push 후 comment 남기도록 하겠습니다. 감사합니다!! (+ READMD.md는 보시기 편하게끔 step 별로 넣어두고 따로 저장해두고 있었습니다. 다음 PR부터 합쳐서 올려두도록 하겠습니다!)

okstring committed:
수정사항 반영

 - rankName 삭제 enum 코드 간소화
 - protocol에 var로 선언해도 let으로 선언 가능
 - CardStud, Participant enum 적극 활용

okstring committed:
appendCard 네이밍 변경, CardBox 클래스 추가, Access Control 재설정, README.md 작성

- getCard -> appendCard
- CardBox 클래스 추가
- Access Control 재 설정(Player.cards) - private는 특정 엔티티가 선언된 괄호 안에서만 사용 가능(엔티티: 데이터의 집합)

okstring commented:
- `rankName` 삭제 enum 코드 간소화
- protocol에 `var`로 선언해도 `let`으로 선언 가능한걸 뒤늦게 알았습니다 😱
- `CardStud`, `Participant` enum 적극 활용
- `getCard` -> `appendCard`
- `CardBox` 클래스 추가
- Access Control 재 설정(`Player.cards`) - `private`는 특정 엔티티가 선언된 괄호 안에서만 사용 가능(엔티티: 데이터의 집합)



#### `playResult`

`playResult`는 다음 스텝부터 요구하는 승패 결과 로직에 따라 적용할 예정이었습니다. 그리고 다시 보니 네이밍도 적절하지 않은 것 같아 `score`로 바꿔봤습니다. 혼선이 생기지 않도록 미처 주석을 달아두지 못했던 것 같습니다. 만약 의도하신 부분이 이부분이 아니라면 다시 한 번 더 생각해보도록 하겠습니다



#### OOP, POP

객체 지향 프로그래밍과 프로토콜 지향 프로그래밍 모두 여러 유형과 상호작용하기 위해 실제 객체 및 Polymorphism을 모델링 하는 것과 같은 철학을 가지고 있습니다. 하지만 이런 철학을 구현하는 방식이 다릅니다. 어느 한쪽이 뛰어나다고 무조건 쓰지 말라고 하는 태도를 지양하는 이유가 실용성의 이유보다 철학의 이유가 크다고 느낍니다. 결국 코드를 잘 짜는 것도 유지보수, 클린코딩의 일환이고 특히 잘 짜여진 코드는 reality를 잘 나타내야 한다고 봅니다. 단적인 예로 수평적인 확장의 느낌을 가지는 프로토콜 지향 프로그래밍만으로 방대한 프로젝트를 꾸몄다고 하면 수직관계 부분은 안보이기 때문에 reality가 떨어진다고 보입니다. 코드를 읽고 수정하면서도 어색한 부분이 많으므로 많은 사람들이 달려드는 프로젝트에도 코드를 읽는 데에 수월하지는 않을 것으로 예상됩니다.

다시 제 코드로 돌아가 모든 카드를 돌려주는 `takeCardSet()`은 `Dealer`의 슈퍼클래스로 해당 관계를 나타내려 코드를 작성했습니다. 그리고 `Playable` protocol의 경우 경기를 참가한 사람들이 가지는 특성에 대해 정의를 하고 묶어놓고 싶어 protocol로 정의하게 됐습니다. 또한 이번 미션에서 두가지 패러다임을 다룬 이유도 많이 헷갈렸던 부분이고 이번 미션에 적용해서 많은 부분 습득할 수 있지 않을까 싶어 두가지 패러다임을 사용하게 됐습니다.

godrm commented:
OOP나 POP를 적용하고 있다니 좋은 시도네요. 
한 가지 힌트를 주자면, OOP든 POP든지 수퍼클래스-서브클래스, 호출하는 상위 모듈-호출당하는 하위 모듈이 전자보다 후자(서브 클래스나 하위 모듈)가 더 구체적이고 더 복잡해야 합니다. 
그래야 상위로 올라갈 수록 간결해지는 구조를 가지게 됩니다. 이걸 기준으로 생각해보세요.

godrm commented on PokerGameApp/PokerGameApp/CardBox.swift:
```diff
@@ -0,0 +1,15 @@
+import Foundation
+
+class CardBox {
+    static func takeCardSet() -> [Card] {
```
전체가 가져간다는 의미로는 복수라는 느낌이 없네요. 
덧붙여서 객체를 생성하는 건 팩토리 메소드라고 부릅니다. 배열에 담아주더라도 생성하는 경우는 다른 의미로 부를 수도 있습니다. 참고하세요.

godrm commented on PokerGameApp/PokerGameApp/CardDeck.swift:
```diff
@@ -4,7 +4,7 @@ struct CardDeck {
     
     init() {
-        cards = Card.createCardSet()
+        cards = Dealer.takeCardSet()
```
객체 관계를 정의하다보다면 Dealer가 CardDeck을 가질지, 지금처럼 CardDeck이 Dealer를 가질지 결정해야 합니다.
Dealer가 CardDeck을 포함하는 게 더 이상적이지 않을까요? 여기 코드는 조금 어색합니다.

godrm commented on PokerGameApp/PokerGameApp/Dealer.swift:
```diff
@@ -2,38 +2,42 @@ import Foundation
     var name: String { get }
-    var playResult: [Card] { get }
-    func getCard(_ card: Card)
+    // 이후에 play 결과를 반환합니다. [Card]가 아닌 값이 반환될 수 있습니다.
+    var finalScore: [Card] { get }
```
주석이 더 헷갈리는 정보를 제공하고 있네요. [Card]가 아닌 값이 반환될 수 있다는 건 무슨 의미일까요?

godrm commented on PokerGameApp/PokerGameApp/Dealer.swift:
```diff
@@ -2,38 +2,42 @@ import Foundation
 }
 
-class Dealer: Playable {
+class Dealer: CardBox, Playable {
```
아.. 딜러가 CardBox를 상속하고 있네요.
상속 관계에 대해 다시 한 번 생각해보면 좋겠습니다. 
OOP에 대해 수업했을 때도 같은 부류로 묶을 때 같은 클래스라고 부를 수 있다고 했습니다.
이건 마치 원숭이-바나나 관계처럼 어색합니다.

godrm commented on PokerGameApp/PokerGameApp/Dealer.swift:
```diff
@@ -2,38 +2,42 @@ import Foundation
-        guard numberOfPlayer >= 1 || numberOfPlayer <= 4 else { return }
-        
+    public func start(numberOfPlayer: Participant, stud: CardStud) {
         let players = registerPlayer(of: numberOfPlayer)
```
players 배열은 속성으로 필요없는건가요? 한번씩 나눠갖고 끝인걸까요? 
그리고보니 Dealer가 players를 모두 포함하는 구조네요. 참조를 하는 게 아니라 포함하는 걸 의도한걸까요?

godrm commented on PokerGameApp/PokerGameApp/Dealer.swift:
```diff
@@ -45,22 +49,23 @@ class Dealer: Playable {
         players.forEach { (player) in
-            print("\(player.name)", player.playResult)
+            print("\(player.name)", player.finalScore)
         }
```
출력하는 함수는 이렇게 딜러처럼 데이터를 다루는 타입에 넣지마세요. 
앱을 만들려고 하는 건데 이런 메소드는 앱에 화면을 표시하는데 사용하지 않는 함수가 됩니다. 
그래서 데이터를 전달하는 구조로 만드는게 좋습니다. 그래야 print가 아니라 다른 뷰에 표시도 가능해집니다.

godrm commented on PokerGameApp/PokerGameApp/Player.swift:
```diff
@@ -2,8 +2,8 @@ import Foundation
 
 class Player: Playable {
```
Dealer와 Player가 모두 Playable 을 채택한 접근은 좋습니다.

godrm commented:
Dealer 타입이 모든 것을 다하고 있어서 가장 큰 혼란을 갖고 있는 것 같습니다. Dealer 보다 상위 개념을 만드는 게 어떨까요? 
예를 들면 PokerGame 같은 타입이 있어서, rule 이나 dealer, player, deck을 모두 갖고 있는 개념은 어떤가요?

okstring commented:
감사합니다!! 객체지향 사고가 쉽게 적용되지 않는 것 같습니다 😅 생각의 전환을 해보고 리뷰주신 것들 곰곰히 생각해서 작성해보겠습니다

okstring committed:
주석, CardBox 메소드명 수정

okstring committed:
CardBox를 Card에 상속받게 변경

okstring committed:
PokerGame 구현

 - PokerGame 안에 CardBox 구현
 - CardDeck을 class로 구현, 속성 cards -> cardDeck으로 이름 변경
 - Model 역할을 위해 ViewController에서 출력하도록 변경

okstring committed:
result 주석 수정

okstring commented:
- PokerGame에서 카드뭉치를 준다는 것이 주체가 모호한것 같아 안에 struct로 CardBox를 만들었습니다.
 - 리뷰주신 것을 보고 전체를 다시 보니 `players` 속성이 필요해보이는 것을 느꼈습니다. 이는 테스트 프린트할때도 필요했습니다.
 - `registerPlayer(of:)` 에서 `dealer`도 플레이어 등록을  위해 `players` 를 만들게 됐습니다. 이후 step 다른 메소드에도 쓰일 수 있을 것 같습니다. 
 - `dealer` 자신을 `player`에 넣고 플레이어 등록, 카드 분배 등 역할을 주어 게임플레이와 동시에 게임 진행을 맡을 수 있도록 실제 딜러 역할과 최대한 비슷하게 구성해보고자 했습니다. 하지만 코드 line이 늘어나는 부분은 아쉽게 생각합니다.

godrm commented on PokerGameApp/PokerGameApp/Card.swift:
```diff
@@ -1,6 +1,6 @@
 import Foundation
 
-class Card: CustomStringConvertible {
+class Card: PokerGame, CustomStringConvertible {
```
엇.. PokerGame은 어떤 객체길래 Card가 상속하는 걸까요?
Card보다 일반화된 개념인가요? 
포함 관계와 상속 관계를 구분하는 게 필요한 것 같습니다.
Card는 카드 한 장을 표현하는(모델링하는) 타입입니다. 카드 한 장이 포커 게임과 같은 수준일까요?

godrm commented:
PokerGame 은 데이터 흐름을 다루는 최상위 모듈(타입)이 되는거지만
그렇다고 상속 관계에서 뿌리가 되어야 하는 것은 아닙니다. 
상속 관계가 필요한 객체끼리만 처리하도록 개선해보세요. 이번 개선은 방향이 좀 잘못됐습니다.

okstring commented:
앗 그렇군요 😭 구분지어 생각해보고 다시 개선해서 push 하겠습니다!

okstring committed:
리뷰 후 수정(PokerGame 수정)

 - start, shareCards, registerPlayer를 PokerGame에서 구현
 - Dealer 추후 점수 계산을 위해 calculateScore() 구현

okstring committed:
start, registerPlayer 메소드 위치 수정

okstring committed:
PokerGame, Dealer 상속관계 해제

okstring commented:
계속해서 생각해보고 있습니다. Dealer가 가지고 있던 메소드들을 PokerGame에 어울리는 메소드들에 한해 다 넘기게 되면 슈퍼메소드는 서브메소드보다 복잡해지는 문제가 발생해서 망설이고 있었습니다. 이런저런 고민을 하다 상속관계를 가지지 않는것도 방법이 될까 싶어 수정해봤습니다.

godrm commented:
수고하셨습니다. 다음 단계를 하면서 Players 타입을 선언해보세요.
